Change benchmarks

rynowak opened the pull request:
/cc @JamesNK @kichalla - separated from #591 to make it easier on all of us :)

Add benchmarks that include some HTTP method matching.

Clean up names and name like-kinded benchmarks alphabetically.

Matcher*Benchmark -> E2E including HTTP method selection
MatcherSelectCandidates*Benchmark -> Focused on just URL path processing

rynowak commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/Candidate.cs:
```diff
@@ -0,0 +1,28 @@
+{
+    // This is not yet fleshed out - consider this part of the
+    // work-in-progress definition of CandidateSet.
+    internal readonly struct Candidate
```
I needed to define a little more infrastructure to merge the benchmarks

rynowak commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherAzureBenchmark.cs:
```diff
@@ -1,20 +1,18 @@
 {
     // Generated from https://github.com/Azure/azure-rest-api-specs
-    public partial class AzureMatcherBenchmark : MatcherBenchmarkBase
+    public partial class MatcherAzureBenchmark : MatcherBenchmarkBase
```
Renaming and changing these so they can sort alphabetically. It makes the console UI for this much better to use.

rynowak commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherAzureBenchmark.cs:
```diff
@@ -1,20 +1,18 @@
 
         private BarebonesMatcher _baseline;
         private Matcher _dfa;
-        private Matcher _route;
```
I'm taking the legacy route implementation out of these big tests, because it's just so slow. It makes the benchmarks really painful to work with and isn't adding much value because it's like 100x slower than the baseline.

rynowak commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherBenchmarkBase.cs:
```diff
@@ -40,7 +40,7 @@ internal static MatcherEndpoint CreateEndpoint(string template, string httpMetho
                 new List<MatchProcessorReference>(),
                 0,
-                EndpointMetadataCollection.Empty,
+                new EndpointMetadataCollection(metadata),
```
Oopsies

rynowak commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherSelectCandidatesAzureBenchmark.cs:
```diff
@@ -0,0 +1,72 @@
+        private const int SampleCount = 100;
+
+        private BarebonesMatcher _baseline;
+        private DfaMatcher _dfa;
```
This is the first of a set of 4 small lower level benchmarks for the DFA algorithm. I'm just testing the DFA here because the other ones don't have a similar layering.

rynowak commented on src/Microsoft.AspNetCore.Routing/Matchers/FastPathTokenizer.cs:
```diff
@@ -9,6 +9,16 @@ namespace Microsoft.AspNetCore.Routing.Matchers
+        //
+        // So setting this limit to 32 should work pretty well. We also expect the tokenizer
+        // to be used with stackalloc, so we want a small number.
+        public const int DefaultSegmentCount = 32;
```
This is from PR feedback I already addressed in the other PR 😆

JamesNK commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherSelectCandidatesAzureBenchmark.cs:
```diff
@@ -0,0 +1,72 @@
+            for (var i = 0; i < SampleCount; i++)
+            {
+                var sample = _samples[i];
+                var httpContext = _requests[sample];
```
Can `_requests` and `_endpoints` be protected instead of internal? Also IMO rename to `Requests` and `Endpoints`. I reserve underscores for only private fields.

rynowak replied:
The can be `private protected`. The reason they weren't is that benchmark classes have to be public, and many of these types are internal.

rynowak replied:
BTW it take a really really long time to rename these due to the generated code 🤣

JamesNK commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherSelectCandidatesSmallEntryCountBenchmark.cs:
```diff
@@ -72,43 +68,31 @@ private Matcher SetupMatcher(MatcherBuilder builder)
 
         [Benchmark(Baseline = true)]
-        public async Task Baseline()
+        public unsafe void Baseline()
```
What is unsafe here?

rynowak replied:
Doesn't stackalloc always require unsafe? I'll see what happens if I remove

rynowak replied:
It turns out you're right btw. You don't need unsafe for stackalloc - you need unsafe to dereference a pointer.

JamesNK commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherSelectCandidatesSmallEntryCountBenchmark.cs:
```diff
@@ -72,43 +68,31 @@ private Matcher SetupMatcher(MatcherBuilder builder)
+            var path = httpContext.Request.Path.Value;
+            var buffer = stackalloc PathSegment[FastPathTokenizer.DefaultSegmentCount];
+            var count = FastPathTokenizer.Tokenize(path, buffer, FastPathTokenizer.DefaultSegmentCount);
+            var segments = new ReadOnlySpan<PathSegment>((void*)buffer, count);
```
Is unsafe code required here? I thought a `stackalloc`ed array implicitly converted to a Span/ReadOnlySpan.

rynowak replied:
You might be right. I may be able to get rid of unsafe in a bunch of places.

JamesNK commented on test/Microsoft.AspNetCore.Routing.Tests/Matchers/BarebonesMatcher.cs:
```diff
@@ -47,27 +47,32 @@ public sealed class InnerMatcher : Matcher
         {
             private readonly string[] _segments;
             public readonly MatcherEndpoint _endpoint;
```
Change to private or rename to Endpoint

JamesNK commented on test/Microsoft.AspNetCore.Routing.Tests/Matchers/DfaMatcher.cs:
```diff
@@ -65,7 +63,15 @@ public unsafe override Task MatchAsync(HttpContext httpContext, IEndpointFeature
+                var httpMethodConstraint = candidate.Endpoint.Metadata.GetMetadata<HttpMethodEndpointConstraint>();
+                if (httpMethodConstraint != null && !MatchHttpMethod(httpContext.Request.Method, httpMethodConstraint))
+                {
+                    goto notmatch;
```
Is this goto really required? Isn't it just skipping one line immediately below it?

rynowak replied:
Dang. I try to hide one goto a week in the product and I failed 😢

JamesNK commented on test/Microsoft.AspNetCore.Routing.Tests/Matchers/DfaMatcher.cs:
```diff
@@ -76,17 +82,39 @@ public unsafe override Task MatchAsync(HttpContext httpContext, IEndpointFeature
+            return states[current].Matches;
+        }
+
+        public struct State
```
copy pasta the readonly `State` from other PR

rynowak commented:
@JamesNK updated

I addressed the feedback and did some general cleanup around unsafe. Using `Span<>` instead of a pointer in `FastPathTokenizer` had no negative perf impact so I made that change, and resolved all of the fallout from it. 

I've deliberately left unsafe and pointers in the baselines for the path tokenizer.
